docs: add migration guide for plugins and service loader (#1429)
kestra-io/docs: content/docs/11.migration-guide/plugins-discover-mechanism.md

---
title: Plugins Discovery Mechanism
icon: /docs/icons/migration-guide.svg
release: 0.17.0
---

> In Kestra version 0.17.0, we decided to redesign the mechanism used to discover and load core and custom plugins.
> 
> Plugins are now discovered and loaded using the standard *Java Service Loader*.

Reviewer: ```suggestion
Kestra 0.17.0 uses a new mechanism to discover and load plugins. If you use custom plugins, follow this guide to make the necessary adjustments.

Plugins are now discovered and loaded using the standard *Java Service Loader*.
```

## Why the Change?

> In earlier versions, Kestra was relying heavily on the _Bean Introspection_ mechanism provided by Micronaut Framework
> for loading plugins (as a reminder Micronaut is the main JVM-based framework used by Kestra).
> 
> However, we have repeatedly encountered some limitations in maintaining backward compatibility of plugins during major
> version upgrades of Micronaut. In addition, this implementation was limiting our ability to introduce future 
> enhancements around the Kestra plugin mechanism.

Reviewer: ```suggestion
So far, Kestra heavily relied on the _Bean Introspection_ mechanism provided by the Micronaut Framework for loading plugins (_Micronaut is the JVM-based API framework used by Kestra_).

However, we have repeatedly encountered some limitations in maintaining the backward compatibility of plugins during major version upgrades of Micronaut. In addition, this implementation was limiting our ability to introduce future enhancements around the plugin mechanism.
```

> Thanks to this new implementation, we have been allowed not only to reduce the number of dependencies required for
> developing custom plugins, but also to divide by two the start-time of Kestra when loading a lot of plugins.

Reviewer: ```suggestion
Thanks to this new implementation, we reduced the number of dependencies required for developing custom plugins, and these plugins now load twice as fast as before.
```

Finally, this change is part of a wider effort to improve the developer experience around plugins, and
to reduce Micronaut's exposure outside the Kestra core.

Unfortunately, we've had to introduce some minor breaking-changes to the way custom plugins should be built.
Here are the changes required to migrate to Kestra 0.17.0.

## Micronaut dependencies

For most plugin implementations, all Micronaut libs can be removed from the `compileOnly` dependencies in the `build.gradle` file.

However, Micronaut is still required to use the utility classes provided by Kestra for running unit-tests.

## Kestra's annotation processor

Kestra requires a new annotation processor to be configured in the `build.gradle` file of your project (or `pom.xml` for Maven).

```
annotationProcessor group: "io.kestra", name: "processor", version: kestraVersion
```

The role of this processor is to automatically manage the `META-INF/services` file needed by Java to discover your plugins.

## Custom Validations

> From the beginning, Kestra allows you to develop custom constraint validator using the standard Java API for bean validation (i.e: JSR-380).
> Custom validation are used to validate the properties of custom tasks.
> 
> Custom validator must now implement the standard `jakarta.validation.ConstraintValidator`
> instead of the interface provided by Micronaut: `io.micronaut.validation.validator.constraints.ConstraintValidator`.
> 
> In addition, custom validation annotation should now strictly adhere to the Java bean specification:
> 
> Here is a migration example:

Reviewer: ```suggestion
Kestra allows you to develop a custom constraint validator using the standard Java API for bean validation (i.e., JSR-380), which is used to validate the properties of custom tasks.

::alert{type="warning"}
The custom validator must now implement the standard `jakarta.validation.ConstraintValidator`
instead of the interface provided by Micronaut: `io.micronaut.validation.validator.constraints.ConstraintValidator`.
::

In addition, custom validation annotation should now strictly adhere to the Java bean specification:

Here is a migration example:
```

> Kestra 0.16 and priors:

Reviewer: ```suggestion
Kestra 0.16.6 and before:
```

```java
// file: io.kestra.plugins.custom.CustomNotEmpty.java
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = CustomNotEmptyValidator.class)
public @interface CustomNotEmpty {
    String message() default "invalid";
}
```

```java
// file: io.kestra.plugins.custom.CustomNotEmptyValidator.java
import io.micronaut.validation.validator.constraints.ConstraintValidator;
import io.micronaut.validation.validator.constraints.ConstraintValidatorContext;
// ...
@Singleton
@Introspected
public class CustomNotEmptyValidator implements ConstraintValidator<CustomNotEmpty, String> {
    @Override
    public boolean isValid(
        @Nullable String value,
        @NonNull AnnotationValue<CustomNotEmpty> annotationMetadata,
        @NonNull ConstraintValidatorContext context) {
        if (value == null) {
            return true; // nulls are allowed according to spec
        } else if (value.size() < 2) {
            context.messageTemplate("string must have at-least two characters");
            return false;
        } else {
            return true;
        }
    }
}
```

> As of Kestra v0.17.

Reviewer: ```suggestion
Kestra 0.17.0 and later:
```
